This pull request is failing during `make` in the `g++` section of our GH CI.  See:
https://github.com/Perl/perl5/actions/runs/22072615631/job/63788157408?pr=24203.

I was able to replicate this build failure on a g++-13 debugging build on Linux.
```
$ g++ --version
g++ (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0

$ sh ./Configure -des -Dusedevel -Dcc=g++ -DDEBUGGING && make test_prep
...
g++ -c -DPERL_CORE -fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -Wall -Werror=pointer-arith -Werror=vla -Wextra -Wno-long-long -Wno-nonnull-compare -Wwrite-strings -Wno-use-after-free perlio.c
In file included from perl.h:6263,
                 from perlio.c:43:
perlio.c: In function ‘ssize_t PerlIOScalar_write(PerlIOl**, const void*, size_t)’:
perlio.c:1271:28: error: invalid conversion from ‘const void*’ to ‘const char*’ [-fpermissive]
 1271 |             vbuf = savepvn(vbuf, count);
      |                            ^~~~
      |                            |
      |                            const void*
embed.h:619:68: note: in definition of macro ‘savepvn’
  619 | # define savepvn(a,b)                           Perl_savepvn(aTHX_ a,b)
      |                                                                    ^
In file included from perl.h:7986:
inline.h:5044:32: note:   initializing argument 1 of ‘char* Perl_savepvn(const char*, size_t)’
 5044 | Perl_savepvn(pTHX_ const char *pv, Size_t len)
      |                    ~~~~~~~~~~~~^~
make: *** [makefile:265: perlio.o] Error 1
```
Above I configured with `-DDEBUGGING` to simulate our CI configuration -- but we get the build failure in non-debugging builds as well.
```
$ sh ./Configure -des -Dusedevel -Dcc=g++ && make test_prep
...
g++ -c -DPERL_CORE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -Wall -Werror=pointer-arith -Werror=vla -Wextra -Wno-long-long -Wno-nonnull-compare -Wwrite-strings -Wno-use-after-free perlio.c
In file included from perl.h:6263,
                 from perlio.c:43:
perlio.c: In function ‘ssize_t PerlIOScalar_write(PerlIOl**, const void*, size_t)’:
perlio.c:1271:28: error: invalid conversion from ‘const void*’ to ‘const char*’ [-fpermissive]
 1271 |             vbuf = savepvn(vbuf, count);
      |                            ^~~~
      |                            |
      |                            const void*
embed.h:619:68: note: in definition of macro ‘savepvn’
  619 | # define savepvn(a,b)                           Perl_savepvn(aTHX_ a,b)
      |                                                                    ^
In file included from perl.h:7986:
inline.h:5044:32: note:   initializing argument 1 of ‘char* Perl_savepvn(const char*, size_t)’
 5044 | Perl_savepvn(pTHX_ const char *pv, Size_t len)
      |                    ~~~~~~~~~~~~^~
make: *** [makefile:265: perlio.o] Error 1
```

--- a/perlio.c
+++ b/perlio.c
@@ -1268,7 +1268,7 @@ PerlIOScalar_write(pTHX_ PerlIO * f, const void *vbuf, Size_t count)
         if (SvTYPE(sv) >= SVt_PV && SvPVX_const(sv) && SvLEN(sv) &&
             UNLIKELY(SvPVX_const(sv) <= (const char *)vbuf &&
                      (const char *)vbuf < SvPVX_const(sv) + SvLEN(sv))) {
-            vbuf = savepvn(vbuf, count);
+            vbuf = savepvn((const char *)vbuf, count);
             SAVEFREEPV(vbuf);
         }
 

--- a/pod/perldelta.pod
+++ b/pod/perldelta.pod
@@ -549,8 +549,9 @@ while being output, garbage bytes would be left in that scalar.
 =item *
 
 Additionally, there was a case that would expose garbage bytes in the
-target scalar when that scalar is output on it (as in
-C<< open $fh, '>', \$str; print $fh $str; >>).  This is also fixed.
+target scalar or even trigger a segmentation fault when that scalar is
+output on it (as in C<< open $fh, '>', \$str; print $fh $str; >>).
+This is also fixed.  [GH #24199]
 
 =back
 

--- a/t/io/scalar.t
+++ b/t/io/scalar.t
@@ -12,7 +12,7 @@ use strict;
 use Fcntl qw(SEEK_SET SEEK_CUR SEEK_END); # Not 0, 1, 2 everywhere.
 use Errno qw(EACCES);
 
-plan(134);
+plan(135);
 
 my $fh;
 my $var = "aaa\n";
@@ -567,4 +567,11 @@ SKIP:
     print $fh $str;
     is($str, "abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz",
        "write a string to itself");
+}
+
+{ # GH #24199
+    open my $fh, '>', \my $content or die $!;
+    print {$fh} "o";
+    print {$fh} $content for 1..20; # This used to crash with SEGV
+    is(length($content), 2 ** 20, "write a string to itself [GH #24199]");
 }

